### PR TITLE
Add auto-formatter

### PR DESCRIPTION
**Pull request by ksivaman**
# Description

Lays out initial infrastructure for auto-formatting/linting the codebase. The key challenge is to find the optimal configs for linters and formatters that do not contradict each other whilst still being useful and increasing productivity. A following PR will enable pre-commit.ci and include all the formatting changes.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Changes

1. Introduces integration with precommit.ci that autoformats the codebase on every PR.
2. Use clang-format for core lib, c++, and cuda files/extensions.
3. Use black for python files.
4. Change linting configs for `cpplint` and `pylint` to be compatible with formatters.
5. Moves some linting functionality such as readability, whitespaces, line lengths etc. from the linters to the formatters such that the 2 do not collide.
6. Consolidates `pylint` and `cpplint` config files to top level.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/distributed.py [hunk 1]
```diff
@@ -615,7 +615,7 @@ def checkpoint(
             "`get_rng_state_tracker` must be passed as keyword arguments to `checkpoint`.",
             DeprecationWarning, stacklevel=2,
         )
-        distribute_saved_activations, get_rng_state_tracker, tp_group = args[:3] # pylint: disable=unbalanced-tuple-unpacking
+        distribute_saved_activations, get_rng_state_tracker, tp_group = args[:3]
```
Comment (timmoon10) [inline transformer_engine/pytorch/distributed.py:618]: I feel like it's better to split up a one-liner than to remove a Pylint warning:
```suggestion
        distribute_saved_activations = args[0]
        get_rng_state_tracker = args[1]
        tp_group = args[2]
```

File: transformer_engine/paddle/utils.py [hunk 2]
```diff
@@ -73,12 +73,12 @@ def mask_to_cu_seqlens(mask: paddle.Tensor, need_kv: bool = False) -> paddle.Ten
     """Convert mask to cu_seqlens"""
     assert 'bool' in str(mask.dtype), "mask must be bool dtype"
     assert len(mask.shape) == 4 and mask.shape[1] == 1, "mask must be [b, 1, s_q, s_kv]"
-    q_actual_seqlens = paddle.sum(mask[:, :, :, 0] == False, axis=(-1, -2), dtype='int32')    # pylint: disable=singleton-comparison
+    q_actual_seqlens = paddle.sum(mask[:, :, :, 0] is False, axis=(-1, -2), dtype='int32')
```
Comment (timmoon10) [inline transformer_engine/paddle/utils.py:76]: I think the `is` will evaluate to a single `False`, so the sum will always be 0.
```suggestion
    q_actual_seqlens = paddle.sum(mask[:, :, :, 0].logical_not(), axis=(-1, -2), dtype='int32')
```

File: transformer_engine/paddle/utils.py [hunk 3]
```diff
@@ -73,12 +73,12 @@ def mask_to_cu_seqlens(mask: paddle.Tensor, need_kv: bool = False) -> paddle.Ten
...
     q_cu_seqlens = paddle.cumsum(q_actual_seqlens)
     q_cu_seqlens = paddle.concat([paddle.zeros([1], dtype=paddle.int32), q_cu_seqlens], axis=0)
     if not need_kv:
         return q_cu_seqlens, None
-    kv_actual_seqlens = paddle.sum(mask[:, :, 0, :] == False, axis=(-1, -2), dtype='int32')    # pylint: disable=singleton-comparison
+    kv_actual_seqlens = paddle.sum(mask[:, :, 0, :] is False, axis=(-1, -2), dtype='int32')
```
Comment (timmoon10) [inline transformer_engine/paddle/utils.py:81]:
```suggestion
    kv_actual_seqlens = paddle.sum(mask[:, :, 0, :].logical_not(), axis=(-1, -2), dtype='int32')
```

File: pylintrc [hunk 4]
```diff
@@ -24,7 +29,10 @@ disable=too-many-locals,
         global-statement,
         too-many-branches,
         global-variable-not-assigned,
-        redefined-argument-from-local
+        redefined-argument-from-local,
+        line-too-long,
+        too-many-return-statements,
+        unbalanced-tuple-unpacking
```
Comment (timmoon10) [inline pylintrc:35]: My thoughts:
- `line-too-long`: One of the biggest stylistic advantages of linting. Strong vote to keep.
- `too-many-return-statements`: I'm fine with simple functions with many return statements. That said, we only need to exclusion once in [`FusedScaleMaskSoftmax.is_kernel_available`](https://github.com/NVIDIA/TransformerEngine/blob/3b813689c16fd5fd286bd2302fc8242afb62b44f/transformer_engine/pytorch/softmax.py#L339).
- `unbalanced-tuple-unpacking`: I don't see a reason to exclude it and it helps with safety.
```suggestion
        redefined-argument-from-local
```

File: CPPLINT.cfg [hunk 5]
```diff
@@ -1,17 +1,19 @@
...
+
+# Ignore the following errors.
+filter=-build/include_subdir
+filter=-build/namespaces
+filter=-readability/todo
+filter=-build/header_guard
+filter=-build/include
+filter=-build/c++11
+filter=-runtime/references
+filter=-whitespace
```
Comment (timmoon10) [inline CPPLINT.cfg:19]: Taming whitespace is one of the big stylistic advantages of linting.
```suggestion
```

File: CPPLINT.cfg [hunk 6]
```diff
@@ -1,17 +1,19 @@
...
+linelength=100
+
+# Ignore the following errors.
+filter=-build/include_subdir
+filter=-build/namespaces
+filter=-readability/todo
+filter=-build/header_guard
+filter=-build/include
+filter=-build/c++11
+filter=-runtime/references
```
Comment (timmoon10) [inline CPPLINT.cfg:18]: I'm fine with this exclusion. I've always found it kind of weird that cpplint doesn't like non-const references since my intuition is that references are safer than pointers.

File: transformer_engine/pytorch/softmax.py [hunk 7]
```diff
@@ -336,7 +336,7 @@ def forward(
             return self.forward_fused_softmax(inp, mask, scale)
         return self.forward_torch_softmax(inp, mask, scale)
 
-    def is_kernel_available(self, mask: torch.Tensor, b: int, np: int, sq: int, sk: int) -> bool: # pylint: disable=too-many-return-statements
+    def is_kernel_available(self, mask: torch.Tensor, b: int, np: int, sq: int, sk: int) -> bool:
```
Comment (timmoon10) [inline transformer_engine/pytorch/softmax.py:339]: I'd prefer manually disabling the Pylint warning here rather than adding an exclusion to the pylintrc:
```suggestion
    def is_kernel_available(self, mask: torch.Tensor, b: int, np: int, sq: int, sk: int) -> bool: # pylint: disable=too-many-return-statements
```

File: .pre-commit-config.yaml
Comment (timmoon10) [inline .pre-commit-config.yaml:1]: We should make sure not to turn on pre-commit right now. It would be better to have a dedicated formatting PR with no functional changes.

**Review by timmoon10 — CHANGES_REQUESTED** [review]

File: CPPLINT.cfg (hunk 6 above)
Comment (ksivaman) [reply CPPLINT.cfg:18]: Google c++ style guide also updated to include this, however cpplint wasn't updated.

File: CPPLINT.cfg (hunk 5 above)
Comment (timmoon10) [reply CPPLINT.cfg:19]: Whitespace will be handled by the formatter, not the linter.

File: pylintrc (hunk 4 above)
Comment (timmoon10) [reply pylintrc]: Line length will be handled by the formatter, not the linter.

**Comment by ksivaman** [comment]
pipeline 15815486

**Review by timmoon10 — APPROVED** [review]

**Review by phu0ngng — APPROVED** [review]
